compiler/optimizer/ValuePropagationCommon.cpp

```diff
@@ -4134,8 +4134,15 @@ void OMR::ValuePropagation::transformArrayCloneCall(TR::TreeTop *callTree, OMR::
       arraycopy->setWordElementArrayCopy(true);
       }
 
+   // Create a new allocation fence ensuring visibility of the newly "cloned" object.
+   TR::Node *allocationFence = TR::Node::createAllocationFence(newArray, newArray);
+
    callTree->insertBefore(TR::TreeTop::create(comp(), TR::Node::create(callNode, TR::treetop, 1, newArray)));
    callTree->insertBefore(TR::TreeTop::create(comp(), TR::Node::create(callNode, TR::treetop, 1, arraycopy)));
+   // Flush after the arraycopy so that the cloned array appears identical to the original before it's made
+   // visible to other threads, and because the obj allocation is allowed to use non-zeroed TLH, we need to
+   // make sure no thread sees stale memory contents from the array element section.
+   callTree->insertBefore(TR::TreeTop::create(comp(), allocationFence));
```

Comment:
Are allocation fence nodes generated in general after allocations on platforms such as X86, that do not have weak memory model ? I ask, because this insertion of an allocation fence seems to be unconditional in that sense, it will add it across platforms. This may not be incorrect, but I thought I'd check if it was unconventional (e.g. if no one code ever generated an allocation fence on non-weak ordered platforms).

Reply:
Yes, good catch.. I had looked at this but somehow missed the `cg()->getEnforceStoreOrder()` check in the genFence() routine used in ILGen. I added that check now.